packages/flutter_tools/lib/src/base/os.dart

```diff
@@ -594,11 +594,11 @@ String? findProjectRoot(FileSystem fileSystem, [ String? directory ]) {
     if (fileSystem.isFileSync(fileSystem.path.join(directory!, kProjectRootSentinel))) {
       return directory;
     }
-    final String parent = fileSystem.path.dirname(directory);
-    if (directory == parent) {
+    final Directory currentDirectory = fileSystem.directory(directory);
+    if (!currentDirectory.existsSync() || currentDirectory.parent.path == currentDirectory.path) {
       return null;
     }
-    directory = parent;
+    directory = currentDirectory.parent.path;
```

Comment:
why don't we just loop over a Directory, rather than casting back and forth from a String <-> Directory? Also, when we instantiate Directory, we should use:

```dart
Directory current = fileSystem.currentDirectory.absolute;
```

Right now this doesn't handle `flutter pub get .` if the current working directory is a sub-directory of the root.

Reply:
This version for me handles `.` in a sub-directory:

```
String? findProjectRoot(FileSystem fileSystem, [ String? directory ]) {
  const String kProjectRootSentinel = 'pubspec.yaml';
  directory ??= fileSystem.currentDirectory.path;
  Directory current = fileSystem.directory(directory).absolute;
  while (true) {
    if (current.childFile(kProjectRootSentinel).existsSync()) {
      return current.path;
    }

    if (!current.existsSync() || current.parent.path == current.path) {
      return null;
    }
    current = current.parent;
  }
}
```